### ISSUE TITLE
Crazy touchpad

### ISSUE THREAD

**Issue by tobiasBora** [NixOS/nixpkgs#118681]
**Describe the bug**

Since I got my laptop (Dell Lattitude 5500) and installed NixOs nearly 1 year ago, my touchpad behave very strangely from time to time (maybe 5mn per hour, but sometimes it works fine during 1 day and suddently gets crazy during 10mn). I thought before that it was a hardware issue, but Dell replaced the touchpad, the battery and the keyboard, and I still have the bug. So I'm thinking that it could be a software issue. If it rings a bell to someone, please let me know!

More precisely: 
1. sometimes the touchpad clicks automatically when I press on key on the keyboard, even if my hand is very far from the keyboard. When this happens, It is very hard to type any text since the cursor jumps sometimes to the mouse position, or the text area loses the focus. Similarly, changing windows with Alt+Tab is quite annoying since sometimes it clicks when I press Tab, which cancels the window's change. When this happens long enough, I can even navigate with only my keyboard, by "clicking" after pressing a random key ^^'
2. sometimes (like right now... :'-(), when I click, instead of clicking, the mouse jumps quite far away from the current position (it could be a few cm away). I also noticed that when this arrives, I can spot a place on the touchpad that I call the "center" (when I click there the mouse does not move), and the further I click from this point, the further the mouse jumps. I managed to record it via xev. Here is a normal click:
```
ButtonPress event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17150910, (428,257), root:(1367,286),
    state 0x10, button 1, same_screen YES

ButtonRelease event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17151090, (428,257), root:(1367,286),
    state 0x110, button 1, same_screen YES
```
and here is this buggy click that moves the cursor:
```
MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161179, (449,221), root:(1388,250),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161186, (455,207), root:(1394,236),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161190, (459,202), root:(1398,231),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161196, (465,192), root:(1404,221),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161200, (467,186), root:(1406,215),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161205, (469,181), root:(1408,210),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161209, (470,177), root:(1409,206),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161214, (474,173), root:(1413,202),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161216, (474,169), root:(1413,198),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161223, (476,165), root:(1415,194),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161225, (476,163), root:(1415,192),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161233, (478,162), root:(1417,191),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161235, (478,160), root:(1417,189),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161242, (480,158), root:(1419,187),
    state 0x10, is_hint 0, same_screen YES

MotionNotify event, serial 40, synthetic NO, window 0x6600001,
    root 0x79b, subw 0x0, time 17161244, (482,158), root:(1421,187),
    state 0x10, is_hint 0, same_screen YES
```

3. I also spotted another bug which is quite different since it is always present (well at the beginning it was never present, but after an upgrade (I think), it stoped working). This time, it is not related with the touchpad but with the buttons of the mini-mouse included inside the keyboard (DELL laptops have a second (unpractical) mouse). When I press the middle button, the press events is fired quite lately, only when I release the button. This is quite annoying in 3D softwares where this button is used to drag/drop the view. To take that `xev` shot, I pressed it during 6 seconds... but the two events (press and release) are received at the exact same time: 
```
ButtonPress event, serial 40, synthetic NO, window 0x4c00001,
    root 0x79b, subw 0x0, time 20303363, (92,116), root:(1031,145),
    state 0x10, button 2, same_screen YES

ButtonRelease event, serial 40, synthetic NO, window 0x4c00001,
    root 0x79b, subw 0x0, time 20303363, (92,116), root:(1031,145),
    state 0x210, button 2, same_screen YES
``` 

I've no idea how to debug this, and it is even harder to fix it given that the bug 1 and 2 happens only a few minutes per hour... But if it rings a bell to someone, please let me know! I also saw that DELL released a firmware update for the touchpad a year ago, but I can't manage to install it since I don't have windows, and when I run it in using the livecd Hirensbootcd, it just tells me the device is not found.

**To Reproduce**
Steps to reproduce the behavior (it is very hard to reproduce):
1. Get a DELL Latitude 5500.
2. Use it during a few hours. At some point you may have this weird click when you press a key on the keyboard or when you click with your mouse.

**Metadata**
I'm not sure if the metadata are very relevant since I got this problem since june 2020. but here is it:

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.93, NixOS, 21.03pre-git (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20201205_a5d85d0`
 - channels(root): `"nixos-unstable-21.03pre267299.5ff6700bb82, nixos-21.03pre267299.5ff6700bb82"`
 - channels(me): `"local-nixos-unstable-21.05pre279012.ad47284f8b0, home-manager, local-nixpkgs-unstable-21.03pre267409.f5daa8ab31f, nixos-20-03-20.03.3324.929768261a3, test, nixos-unstable-21.03pre250162.34ad166a830"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by tobiasBora** [comment]
I still have the issue… but I strongly suspect a hardware issue (technically Dell replaced half of my computer, it solved it for like 3 days and after it was back). Dell seems to be really bad on touchpad, I got another Dell (different model), and the touchpad clearly has bugs (it clicks automatically physically with a sound 10x in a row), the bug was different (ubuntu). I guess we can close it… for now my trick is to put the mouse in some corners where clicks are not taken into account ^^'

**Comment by tobiasBora** [comment]
This might actually be related to sleep mode, see this discussion https://gitlab.freedesktop.org/libinput/libinput/-/issues/36 with a fix pushed to libinput 1.12 (I still see some weird touchpad related stuff sometimes, so not sure if it's 100% solved… but it does occur less often than before and in a less dramatic way as the cursor stopped from crossing the whole screen)
